jurity/recommenders/auc.py

```diff
@@ -0,0 +1,155 @@
+            multiple users per DataFrame. However, the interactions for a specific user must be contained within a
+            single DataFrame.
+        predicted_results: pd.DataFrame
+            A pandas DataFrame for the recommended user item interaction data, captured from a recommendation algorithm.
+            The DataFrame should contain a minimum of two columns, including self._user_id_column, self._item_id_column,
+            and anything else the metric may need. Each row contains the interaction of one user with one item, and the
+            scores associated with this interaction. There can be multiple interactions per user, and there can be
+            multiple users per DataFrame. However, the interactions for a specific user must be contained within a
+            single DataFrame.
+        batch_accumulate: bool
+            If specified, this parameter allows you to pass in minibatches of results and accumulate the metric
+            correctly across the batches. This reduces the memory footprint and integrates easily with batched
+            training. If specified, the ``get_score`` function will return a tuple of batch results and accumulated
+            results.
+        return_extended_results: bool
+            Whether the extended results such as the support should also be returned. If specified, the returned results
+            will be of type ``dict``. AUC currently returns ``auc`` and the ``support`` used to calculate AUC.
+
+        Returns
+        -------
+        metric: Union[float, dict, Tuple[float, float], Tuple[dict, dict]]
+            The averaged result(s). The return type is determined by the ``batch_accumulate`` and
+            ``return_extended_results`` parameters. See the examples above.
+        """
+        actual_results = actual_results.set_index([self._user_id_column, self._item_id_column])
+        predicted_results = predicted_results.set_index([self._user_id_column, self._item_id_column])
+        if self.k is not None:
+            sorted_clicks = get_sorted_clicks(predicted_results, self._user_id_column, self.click_column, self.k)
+        else:
+            sorted_clicks = predicted_results
+
+        matches = actual_results.join(sorted_clicks, how='inner', rsuffix='_r')
+        clicks = matches[self.click_column].values
+        predictions = matches[self.click_column + '_r'].values
+        results = np.vstack((clicks, predictions)).T
+
+        return self._accumulate_and_return(results, batch_accumulate, return_extended_results)
+
+    @staticmethod
+    def _get_results(results: List[np.ndarray]) -> float:
+        results = np.concatenate(results)
+        return roc_auc_score(results[:, 0], results[:, 1])
+
+    def _get_extended_results(self, results: List[np.ndarray]) -> dict:
+        auc = self._get_results(results)
+        return {'auc': auc, 'support': len(np.concatenate(results))}
```

Comment:
I am struggling with the dimensionality here a bit. Support is the "user-item pairs" matching interactions between predicted and actual... hence, shouldn't this just be len(results) since we previously do the matching? doing np.concatenate (with default axis=0), we would essentially double-count the hits?

Reply:
`results` is a list of numpy array(s) so the concat is concatenating the list(s) not the np array. Might be more clear if you look at the `_accumulate_and_return` implementation in base.py.

Reply:
concatenating list of np.arrays or two numpy arrays should have similar default behaviour, as least as far as I can tell.

![Screen Shot 2021-04-13 at 3 24 24 PM](https://user-images.githubusercontent.com/31483888/114609063-5ddb8280-9c6c-11eb-82f3-fb67c171c89b.png)

I think the reason this works is because of the matrix get transposed and thus the two numpy arrays get appended as columns next to each other. so `len(results[0]` would be the same as `len(np.concatenate(results)` here. I get it now.